New Nuclear explosion type

B-Kirill opened the pull request:
## About the PR
- Adds new `Nuclear` explosion type, which, in addition to the usual explosion damage, causes radiation damage. It also has flags for configuring the removal of some objects (for more information, see the technical part of the PR).
- Changes nuke explosion type from `Default` to `Nuclear`.

## Why / Balance
After a nuclear explosion, a lot of objects remain, including a bunch of items, pipes, wires. Of course, this greatly damages performance (both on the server and on the client), especially when there are a lot of players. 
The new type of explosion not only allows you to remove unnecessary objects during an explosion without affecting other sources of explosions, but also makes it possible to modify the nuke and its explosion in the future, as well as add radiation damage for realism.

## Technical details
`Nuclear` explosion has the following flags:
-  `deleteTags: ["DisposalTube", "Item", "Cable", "WallMount","SpreaderIgnore"]`
-  `blacklistedTags: ["HighRiskItem"]`

Deletion works by checking the desired tag for the item and a `HashSet` ensures no duplicate deletions. Of course you can add other tags. (`SpreaderIgnore` tag is needed to remove small and directional fans.)
Since the objects I needed to delete initially didn't have the tags I needed, I had to create a `MapInitTagSystem.cs` that adds a tag to an object at an event of the same name (which is a fairly optimized way compared to others).

`Nuclear` explosion `damagePerIntensity`:
- `Heat: 5`
- `Blunt: 5`
- `Piercing: 5`
- `Radiation: 1`

## Media
Just for example, let's take the explosion at the Box station:

Before the `Nuclear` explosion was added
![2025-03-24_22-26-46](https://github.com/user-attachments/assets/bb72e8d3-ed78-4746-815d-6d3658b7f772)

After adding the `Nuclear` explosion (Slightly outdated, high risk items are not deleted.)
![2025-03-24_22-18-54](https://github.com/user-attachments/assets/5a31a056-c965-4cd5-b6f2-cb10bc00adac)

## Notes  
This is my first PR to SS14, so any feedback greatly appreciated!

## Requirements
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
:cl:
- add: New Nuclear explosion that deletes cables, pipes, and other objects.

ScarKy0 commented:
Could probably be added to The Nuclear Option as well

NoElkaTheGod commented:
Looks very good, except for the changelog. Technical information is not needed, you may remove it and leave something like "Nuclear explosions now delete cables, pipes, and other obects."

B-Kirill commented on Content.Server/Explosion/EntitySystems/ExplosionSystem.Processing.cs:
```diff
@@ -466,6 +473,45 @@ private void ProcessEntity(
 
+                if (type.DeleteItems)
+                {
+                    if (_entMan.EntityExists(entity) && !_processedForDeletion.Contains(entity) && !_entMan.IsQueuedForDeletion(entity) && _itemQuery.HasComponent(entity) && !_container.IsEntityInContainer(entity))
+                    {
+                        //Logger.Info($"Deleting item: {ToPrettyString(entity)} | DeleteItems: {type.DeleteItems} | Type: {type.ID}");
+                        _processedForDeletion.Add(entity);
+                        _entMan.QueueDeleteEntity(entity);
+                    }
+                }
+
+                if (type.DeleteConstructions)
+                {
+                    if (_entMan.EntityExists(entity) && !_processedForDeletion.Contains(entity) && !_entMan.IsQueuedForDeletion(entity) && _constructionQuery.HasComponent(entity))
+                    {
+                        //Logger.Info($"Deleting construction: {ToPrettyString(entity)} | DeleteConstructions: {type.DeleteConstructions} | Type: {type.ID}");
+                        _processedForDeletion.Add(entity);
+                        _entMan.QueueDeleteEntity(entity);
+                    }
+                }
+
+                if (type.DeleteWallMounts)
+                {
+                    if (_entMan.EntityExists(entity) && !_processedForDeletion.Contains(entity) && !_entMan.IsQueuedForDeletion(entity) && _wallMountQuery.HasComponent(entity))
+                    {
+                        //Logger.Info($"Deleting wallmount: {ToPrettyString(entity)} | DeleteWallMounts: {type.DeleteWallMounts} | Type: {type.ID}");
+                        _processedForDeletion.Add(entity);
+                        _entMan.QueueDeleteEntity(entity);
+                    }
+                }
+
+                if (type.DeleteCables)
+                {
+                    if (_entMan.EntityExists(entity) && !_processedForDeletion.Contains(entity) && !_entMan.IsQueuedForDeletion(entity) && _cableQuery.HasComponent(entity))
+                    {
+                        //Logger.Info($"Deleting cable: {ToPrettyString(entity)} | DeleteCables: {type.DeleteCables} | Type: {type.ID}");
+                        _processedForDeletion.Add(entity);
+                        _entMan.QueueDeleteEntity(entity);
+                    }
+                }
```
If it's all about optimization, I can rewrite this block of code for better performance by reducing the number of checks. But about the hardcode of the components, I do not know a better implementation method, even in theory.

B-Kirill replied:
Fixed it in commit edc3580 (forgot to tell you). Now objects are deleted by having a tag, the list of which is set in the explosion settings.

Everturning commented:
Gone. Reduced to atoms.

Pixeltheaertist commented:
Are you still protected inside of Fridges?

RedBookcase commented:
> Are you still protected inside of Fridges?

Fridges had their nuke explosion resistance removed because it was causing unintended fun.

B-Kirill commented:
Well, I've updated the PR description, now I'll deal with the problem..

slarticodefast commented on Content.Server/Explosion/EntitySystems/ExplosionSystem.Processing.cs:
```diff
@@ -463,9 +470,44 @@ private void ProcessEntity(
 
-                // TODO EXPLOSIONS turn explosions into entities, and pass the the entity in as the damage origin.
-                _damageableSystem.TryChangeDamage(entity, damage * _damageableSystem.UniversalExplosionDamageModifier, ignoreResistances: true);
+                if (type.BlacklistedTags != null)
+                {
+                    bool skipDeletion = false;
+                    foreach (var blacklistTag in type.BlacklistedTags)
+                    {
+                        if (_tagSystem.HasTag(entity, blacklistTag))
+                        {
+                            //Logger.Info($"Skipped deletion for {ToPrettyString(entity)} due to blacklisted tag: {blacklistTag}");
+                            skipDeletion = true;
+                            break;
+                        }
+                    }
+                        if (skipDeletion)
+                        continue;
+                }
```
This is basically an `EntityWhitelist` with extra steps. Not sure what the performance impact would be here.

B-Kirill replied:
In my tests, I noticed almost no performance difference between a `Default` and a `Nuclear` explosion. It even seemed to me that it was better, although who knows my PC.
Still, I don't know a better way, but for now I'll deal with other comments.

metalgearsloth replied:
Did you use dottrace API? You can't just look at the FPS counter.

B-Kirill replied:
Well, fair enough...

slarticodefast commented on Content.Server/Tags/MapInitTagSystem.cs:
```diff
@@ -0,0 +1,45 @@
+using Robust.Shared.GameObjects;
+
+namespace Content.Server.Tag;
+public sealed class MapInitTagSystem : EntitySystem
```
This whole system is not needed since you can just directly add a tag to any protoype with the `TagComponent`.
Also it does not seem to be used anywhere as the component and subscription are missing.

B-Kirill replied:
@slarticodefast 
I apologize for the inconvenience — I wanted to know if you received my previous question about tags?

slarticodefast replied:
Hmm, I get your point. But I think
- having two different methods to add tags to objects that are inherited differently will make yaml inheritance a total mess over time, especially since you won't be able to see one of them in the prototype.
- if we have to add a tag to nearly every entity in the game then there is something fundamentally wrong with the code design or api tooling.
- Like sloth said, checking every single entity caught in the  explosion is really expensive

slarticodefast commented on Content.Shared/Explosion/ExplosionPrototype.cs:
```diff
@@ -110,6 +110,12 @@ public sealed partial class ExplosionPrototype : IPrototype
     [DataField("fireStates")]
     public int FireStates = 3;
 
+    [DataField("deleteTags")]
+    public HashSet<string> DeleteTags { get; } = new();
+
+    [DataField("blacklistedTags")]
+    public HashSet<string> BlacklistedTags { get; } = new();
```
Getters and setters for datafields are no longer needed.
Also just use `[DataField]`, the name gets automatically converted to camelCase.

B-Kirill replied:
```
[DataField]
public HashSet<ProtoId<TagPrototype>> deleteTags = new();

[DataField]
public HashSet<ProtoId<TagPrototype>> blacklistedTags = new();
```
Something like that?

slarticodefast replied:
yep, but like sloth mentionend, using tags is bad for performance

slarticodefast reviewed (changes requested):
Again, I'm not sure about the performance impact when checking a tag for every single item caught in a station-wide explosion. Maybe another maintainer can chime in here, as I cannot think of any elegant solutions without investigating the code more.

metalgearsloth commented on Content.Server/Explosion/EntitySystems/ExplosionSystem.Processing.cs:
```diff
@@ -28,6 +28,8 @@ namespace Content.Server.Explosion.EntitySystems;
 public sealed partial class ExplosionSystem
 {
     [Dependency] private readonly FlammableSystem _flammableSystem = default!;
+    [Dependency] private readonly IEntityManager _entMan = default!;
+    private HashSet<EntityUid> _processedForDeletion = new();
```
EntityManager is already a dependency on systems, it's why they're systems.

metalgearsloth commented on Content.Server/Explosion/EntitySystems/ExplosionSystem.Processing.cs:
```diff
@@ -463,9 +470,44 @@ private void ProcessEntity(
+                        continue;
+                }
 
+                if (type.DeleteTags.Count > 0 && 
+                    _tagQuery.TryGetComponent(entity, out var tags))
+                {
+                    foreach (var tag in type.DeleteTags)
+                    {
+                        if (_tagSystem.HasTag(entity, tag) && 
+                            !_processedForDeletion.Contains(entity) && 
+                            !_entMan.IsQueuedForDeletion(entity) && 
+                            !_container.IsEntityInContainer(entity))
+                        {
+                            //Logger.Info($"Deleting: {ToPrettyString(entity)} | Type: {type.ID} | Tag: {tag}");
+                            _processedForDeletion.Add(entity);
+                            _entMan.QueueDeleteEntity(entity);
+                            break;
+                        }
+                    }
+                }
+                if (!_entMan.Deleted(entity) && !_entMan.IsQueuedForDeletion(entity) && !_processedForDeletion.Contains(entity))
+                {
+                    // TODO EXPLOSIONS turn explosions into entities, and pass the the entity in as the damage origin.
+                    _damageableSystem.TryChangeDamage(entity, damage * _damageableSystem.UniversalExplosionDamageModifier, ignoreResistances: true);
+                }
```
Checking for tags and container is really expensive, plus you're not even re-using the components here.

Using damage thresholds is significantly more convenient code-wise and performant just tooling it is bad.

B-Kirill commented:
@slarticodefast 
Hi again!
(I hope I'm not annoying you yet... ` ¯\_(ツ)_/¯`)

**Here’s a breakdown of the changes I’ve made (key points only):**

- Feedback fixes: Addressed all your (and not only yours) recommendations (deleted unnecessary dependencies, adjusted code structure, etc.).

- Optimizations: Squeezed out every drop of efficiency possible:

  - Minimized tag checks.

  - Optimized hashing for faster lookups.

  - Skipped redundant operations where possible.

  - Default explosions: My code now just skips processing them. (Not sure why I didn’t think of this earlier…)

**Performance concerns:**
Your worries about heavy performance impact might’ve been a bit overblown (_or maybe I just optimized it well, who knows_). I took performance measurements and here are the results:

- Default explosions: My code adds 0–0.0001ms to `ProcessEntity`.

- Nuclear explosions: Ranges from 0.0002ms to 0.2016ms in rare cases (still ≤20% of `ProcessEntity` time).

**Context:** Nukes don’t detonate every round, and the server performance gain from deleting debris far outweighs these costs. This happens at the end of the round, and the explosion of the nuclear bomb itself causes lags, so my code won't change much. - I’d call this a win.

**The Tags Problem:**

Agreed: MapInitTagSystem.cs feels hacky and confusing. But here’s the catch:

- Items: `BaseItem` tags get overridden by child entities. Adding `Item` tag to every child at .yml manually isn’t feasible (there are hundreds).

- WallMounts: No **BaseWallMount** equivalent exists, and adding tags to all of them would be a nightmare.

- Cables and DisposalTubes (instead of Constructions) - in this case, I can add tags to .yml manually, as there are few of them. However, I'm not doing this yet, in case there's a normal way to add tags.


Yes, I still insist on using tags. Everything turned out to be not so bad with their performance, and their convenience in adding to the deletion list covers everything else. It remains only to deal with their adequate addition... But here I can only count on your advice.

Looking forward to your thoughts!

B-Kirill commented:
@slarticodefast 
Hi! Just checking in - want to make sure this PR isn't lost in the shuffle. Please let me know if you need anything from my side.
